File: src/main/java/mediathek/tool/table/MVDownloadsTable.java
package mediathek.tool.table;

import mediathek.config.Daten;
import mediathek.config.MVConfig;
import mediathek.daten.DatenDownload;
import mediathek.gui.messages.DownloadQueueRankChangedEvent;
import mediathek.tool.MessageBus;
import mediathek.tool.models.TModelDownload;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

import javax.activation.DataHandler;
import javax.swing.*;
import java.awt.*;
import java.awt.datatransfer.DataFlavor;
import java.awt.datatransfer.Transferable;
import java.awt.dnd.DragSource;
import java.awt.event.MouseEvent;
import java.util.ArrayList;
import java.util.Optional;

public class MVDownloadsTable extends MVTable {
    public MVDownloadsTable() {
        super(DatenDownload.MAX_ELEM, DatenDownload.spaltenAnzeigen,
                Optional.of(MVConfig.Configs.SYSTEM_TAB_DOWNLOAD_ICON_ANZEIGEN),
                Optional.of(MVConfig.Configs.SYSTEM_TAB_DOWNLOAD_ICON_KLEIN),
                Optional.of(MVConfig.Configs.SYSTEM_EIGENSCHAFTEN_TABELLE_DOWNLOADS));

        setupDragnDrop();

        setModel(new TModelDownload());
    }

    @Override public String getToolTipText(MouseEvent e) {
        var p = e.getPoint(); // MouseEvent
        final int viewColumn = columnAtPoint(p);
        final int modelColumnIndex = convertColumnIndexToModel(viewColumn);

        //only show title as tooltip for TITEL column...
        if (modelColumnIndex != DatenDownload.DOWNLOAD_TITEL)
            return super.getToolTipText(e);

        String toolTipText = null;
        final int viewRow = rowAtPoint(p);
        var comp = prepareRenderer(getCellRenderer(viewRow, viewColumn), viewRow, viewColumn);
        var bounds = getCellRect(viewRow, viewColumn, false);


        try {
            //comment row, exclude heading
            if (comp.getPreferredSize().width > bounds.width) {
                final int modelRowIndex = convertRowIndexToModel(viewRow);
                final DatenDownload datenDownload = (DatenDownload) getModel().getValueAt(modelRowIndex, DatenDownload.DOWNLOAD_REF);
                if (datenDownload.film != null)
                    toolTipText = datenDownload.film.getTitle();
                else
                    toolTipText = "";
            }
        } catch (RuntimeException ignored) {
            //catch null pointer exception if mouse is over an empty line
        }

        return toolTipText;
    }

    private void setupDragnDrop() {
        setDragEnabled(true);
        setDropMode(DropMode.INSERT_ROWS);
        setTransferHandler(new TableRowTransferHandlerDownload(this));
    }

    @Override
    public void resetTabelle() {
        for (int i = 0; i < maxSpalten; ++i) {
            resetDownloadsTab(i);
        }

        super.resetTabelle();
    }

    private void resetDownloadsTab(int i) {
        reihe[i] = i;
        breite[i] = 200;
        switch (i) {
            case DatenDownload.DOWNLOAD_NR, DatenDownload.DOWNLOAD_FILM_NR -> breite[i] = 75;
            case DatenDownload.DOWNLOAD_BUTTON_START, DatenDownload.DOWNLOAD_BUTTON_DEL,
                    DatenDownload.DOWNLOAD_PROGRAMM_RESTART, DatenDownload.DOWNLOAD_PROGRAMM_DOWNLOADMANAGER,
                    DatenDownload.DOWNLOAD_UNTERBROCHEN, DatenDownload.DOWNLOAD_SPOTLIGHT,
                    DatenDownload.DOWNLOAD_SUBTITLE, DatenDownload.DOWNLOAD_INFODATEI,
                    DatenDownload.DOWNLOAD_HD, DatenDownload.DOWNLOAD_UT -> breite[i] = 50;
            case DatenDownload.DOWNLOAD_TITEL -> breite[i] = 250;
            case DatenDownload.DOWNLOAD_ABO, DatenDownload.DOWNLOAD_THEMA -> breite[i] = 150;
            case DatenDownload.DOWNLOAD_DATUM, DatenDownload.DOWNLOAD_ZEIT, DatenDownload.DOWNLOAD_GROESSE,
                    DatenDownload.DOWNLOAD_BANDBREITE, DatenDownload.DOWNLOAD_SENDER, DatenDownload.DOWNLOAD_PROGRESS,
                    DatenDownload.DOWNLOAD_RESTZEIT, DatenDownload.DOWNLOAD_DAUER,
                    DatenDownload.DOWNLOAD_GEO -> breite[i] = 100;
        }
    }

    /**
     * Don´t know exactly why this is actually called or needed...
     * but it sorts the list of downloads to be the same as the view order of this table.
     */
    public synchronized void sortDownloadListByTableRows()
    {
        final var rowCount = getRowCount();
        final var tableModel = getModel();
        final var listeDownloads = Daten.getInstance().getListeDownloads();

        for (int i = 0; i < rowCount; ++i) {
            DatenDownload datenDownload = (DatenDownload) tableModel.getValueAt(convertRowIndexToModel(i), DatenDownload.DOWNLOAD_REF);
            listeDownloads.remove(datenDownload);
            listeDownloads.add(datenDownload);
        }
    }

    @Override
    protected void spaltenAusschalten() {
        for (int i = 0; i < maxSpalten; ++i) {
            switch (i) {
                case DatenDownload.DOWNLOAD_FILM_URL, DatenDownload.DOWNLOAD_URL_RTMP,
                        DatenDownload.DOWNLOAD_URL_SUBTITLE, DatenDownload.DOWNLOAD_PROGRAMM,
                        DatenDownload.DOWNLOAD_PROGRAMM_AUFRUF, DatenDownload.DOWNLOAD_PROGRAMM_AUFRUF_ARRAY,
                        DatenDownload.DOWNLOAD_PROGRAMM_RESTART, DatenDownload.DOWNLOAD_PROGRAMM_DOWNLOADMANAGER,
                        DatenDownload.DOWNLOAD_ZIEL_DATEINAME, DatenDownload.DOWNLOAD_ZIEL_PFAD,
                        DatenDownload.DOWNLOAD_ART, DatenDownload.DOWNLOAD_QUELLE, DatenDownload.DOWNLOAD_ZURUECKGESTELLT,
                        DatenDownload.DOWNLOAD_HISTORY_URL, DatenDownload.DOWNLOAD_REF, DatenDownload.DOWNLOAD_SPOTLIGHT,
                        DatenDownload.DOWNLOAD_INFODATEI, DatenDownload.DOWNLOAD_SUBTITLE,
                        DatenDownload.DOWNLOAD_UNTERBROCHEN -> breite[i] = 0;
            }
        }
    }

    private class TableRowTransferHandlerDownload extends TransferHandler {
        private final DataFlavor localObjectFlavor = new DataFlavor(Integer.class, "Integer Row Index");
        private final JTable table;
        private int[] transferedRows;

        public TableRowTransferHandlerDownload(JTable table) {
            this.table = table;
        }

        @Override
        protected Transferable createTransferable(JComponent c) {
            assert (c.equals(table));
            transferedRows = table.getSelectedRows();
            return new DataHandler(table.getSelectedRow(), localObjectFlavor.getMimeType());
        }

        @Override
        public boolean canImport(TransferHandler.TransferSupport info) {
            try {
                boolean b = info.getComponent() == table && info.isDrop() && info.isDataFlavorSupported(localObjectFlavor);
                table.setCursor(b ? DragSource.DefaultMoveDrop : DragSource.DefaultMoveNoDrop);
                return b;
                // here's the problem
                // canImport is called during drags AND before drop is accepted
            } catch (Exception e) {
                logger.error("canImport", e);
            }
            return true;
        }

        private static final Logger logger = LogManager.getLogger();

        @Override
        public int getSourceActions(JComponent c) {
            return TransferHandler.COPY_OR_MOVE;
        }

        @Override
        public boolean importData(TransferHandler.TransferSupport info) {
            try {
                JTable target = (JTable) info.getComponent();
                JTable.DropLocation dl = (JTable.DropLocation) info.getDropLocation();
                int index = dl.getRow();
                int max = table.getModel().getRowCount();
                if (index < 0 || index > max) {
                    index = max;
                }
                target.setCursor(Cursor.getPredefinedCursor(Cursor.DEFAULT_CURSOR));
                if (transferedRows != null) {
                    reorder(index, transferedRows);
                    transferedRows = null;
                    return true;
                }
            } catch (Exception ex) {
                logger.error("importData", ex);
            }
            return false;
        }

        private void reorder(int index, int[] rowFrom) {
            saveSelectedTableRows();

            final var daten = Daten.getInstance();
            final var tModel = (TModelDownload) getModel();
            // listeDownloads neu nach der Reihenfolge in der Tabelle erstellen
            sortDownloadListByTableRows();

            // Downloads zum Verschieben suchen
            ArrayList<DatenDownload> liste = new ArrayList<>();
            for (int row : rowFrom) {
                if (index > row) {
                    --index;
                }
                final DatenDownload d = ((DatenDownload) tModel.getValueAt(convertRowIndexToModel(row), DatenDownload.DOWNLOAD_REF));
                liste.add(d);
                daten.getListeDownloads().remove(d);
            }

            // an der richtigen Stelle einfügen
            daten.getListeDownloads().addAll(index, liste);
            // die Tabellensortierung löschen, die wird jetzt mit der Liste wieder gefüllt
            getRowSorter().setSortKeys(null);
            setRowSorter(null);
            setAutoCreateRowSorter(true);
            restoreSelectedTableRows();

            MessageBus.getMessageBus().publishAsync(new DownloadQueueRankChangedEvent());
        }

        @Override
        protected void exportDone(JComponent c, Transferable t, int act) {
            if (act == TransferHandler.MOVE) {
                table.setCursor(Cursor.getPredefinedCursor(Cursor.DEFAULT_CURSOR));
            }
        }
    }
}

File: src/main/kotlin/mediathek/tool/table/MVDownloadsTable.kt
/*
 * Copyright (c) 2026 derreisende77.
 * This code was developed as part of the MediathekView project https://github.com/mediathekview/MediathekView
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package mediathek.tool.table

import mediathek.audiothek.ui.table.TriStateTableRowSorter
import mediathek.controller.DownloadColumn
import mediathek.controller.starter.DownloadServices
import mediathek.daten.DatenDownload
import mediathek.tool.models.TModelDownload
import org.apache.logging.log4j.LogManager
import java.awt.Cursor
import java.awt.datatransfer.DataFlavor
import java.awt.datatransfer.Transferable
import java.awt.dnd.DragSource
import java.awt.event.MouseEvent
import javax.activation.DataHandler
import javax.swing.DropMode
import javax.swing.JComponent
import javax.swing.JTable
import javax.swing.JTable.DropLocation as TableDropLocation
import javax.swing.TransferHandler
import javax.swing.table.TableModel

private val logger = LogManager.getLogger()

class MVDownloadsTable(
    private val downloads: DownloadServices,
) : PersistentColumnConfigurationTable(
    DownloadColumn.COUNT,
    DownloadColumn.visibilityStore(),
    TableConfigurationStores.DOWNLOAD,
) {
    private var sorter: DownloadsRowSorter? = null

    init {
        autoCreateRowSorter = false
        addPropertyChangeListener("model") { event ->
            val newModel = event.newValue as? TableModel ?: return@addPropertyChangeListener
            val currentSorter = sorter
            if (currentSorter == null) {
                val createdSorter = DownloadsRowSorter(newModel)
                sorter = createdSorter
                rowSorter = createdSorter
            } else {
                currentSorter.model = newModel
            }
        }
        setupDragAndDrop()
        model = TModelDownload()
    }

    override fun getToolTipText(event: MouseEvent): String? {
        val point = event.point
        val viewColumn = columnAtPoint(point)
        if (viewColumn < 0) {
            return super.getToolTipText(event)
        }
        val modelColumn = DownloadColumn.fromIndex(convertColumnIndexToModel(viewColumn))

        if (modelColumn != DownloadColumn.TITLE) {
            return super.getToolTipText(event)
        }

        val viewRow = rowAtPoint(point)
        val component = prepareRenderer(getCellRenderer(viewRow, viewColumn), viewRow, viewColumn)
        val bounds = getCellRect(viewRow, viewColumn, false)

        return try {
            if (component.preferredSize.width > bounds.width) {
                val modelRowIndex = convertRowIndexToModel(viewRow)
                val download = model.getValueAt(modelRowIndex, DownloadColumn.REF.index) as DatenDownload
                download.film?.title.orEmpty()
            } else {
                null
            }
        } catch (_: RuntimeException) {
            null
        }
    }

    private fun setupDragAndDrop() {
        dragEnabled = true
        dropMode = DropMode.INSERT_ROWS
        transferHandler = TableRowTransferHandlerDownload(this)
    }

    override fun resetTabelle() {
        for (column in 0 until maxSpalten) {
            resetDownloadsTab(column)
        }

        listeSortKeys = null

        rowSorter?.sortKeys = null
        spaltenAusschalten()
        setSpaltenEinAus(breite)
        setSpalten()
        calculateRowHeight()
    }

    private fun resetDownloadsTab(column: Int) {
        reihe[column] = column
        breite[column] = when (DownloadColumn.fromIndex(column)) {
            DownloadColumn.NUMBER,
            DownloadColumn.FILM_NUMBER,
                -> 75

            DownloadColumn.BUTTON_START,
            DownloadColumn.BUTTON_DELETE,
            DownloadColumn.PROGRAM_RESTART,
            DownloadColumn.DOWNLOAD_MANAGER,
            DownloadColumn.INTERRUPTED,
            DownloadColumn.SPOTLIGHT,
            DownloadColumn.SUBTITLE,
            DownloadColumn.INFO_FILE,
            DownloadColumn.HIGH_QUALITY,
            DownloadColumn.SUBTITLE_AVAILABLE,
                -> 50

            DownloadColumn.TITLE -> 250

            DownloadColumn.ABO,
            DownloadColumn.TOPIC,
                -> 150

            DownloadColumn.DATE,
            DownloadColumn.TIME,
            DownloadColumn.SIZE,
            DownloadColumn.BANDWIDTH,
            DownloadColumn.SENDER,
            DownloadColumn.PROGRESS,
            DownloadColumn.REMAINING_TIME,
            DownloadColumn.DURATION,
            DownloadColumn.GEO,
                -> 100

            else -> 200
        }
    }

    /**
     * Don't know exactly why this is actually called or needed, but it sorts the list of downloads
     * to be the same as the view order of this table.
     */
    @Synchronized
    fun sortDownloadListByTableRows() {
        val tableModel = model
        val downloadsInTableOrder = ArrayList<DatenDownload>()

        for (row in 0 until rowCount) {
            val download = tableModel.getValueAt(convertRowIndexToModel(row), DownloadColumn.REF.index) as DatenDownload
            downloadsInTableOrder.add(download)
        }
        downloads.reorderQueueToMatch(downloadsInTableOrder)
    }

    override fun spaltenAusschalten() {
        for (column in 0 until maxSpalten) {
            when (DownloadColumn.fromIndex(column)) {
                DownloadColumn.FILM_URL,
                DownloadColumn.RTMP_URL,
                DownloadColumn.SUBTITLE_URL,
                DownloadColumn.PROGRAM,
                DownloadColumn.PROGRAM_INVOCATION,
                DownloadColumn.PROGRAM_INVOCATION_ARRAY,
                DownloadColumn.PROGRAM_RESTART,
                DownloadColumn.DOWNLOAD_MANAGER,
                DownloadColumn.TARGET_FILE_NAME,
                DownloadColumn.TARGET_PATH,
                DownloadColumn.TYPE,
                DownloadColumn.SOURCE,
                DownloadColumn.DEFERRED,
                DownloadColumn.HISTORY_URL,
                DownloadColumn.REF,
                DownloadColumn.SPOTLIGHT,
                DownloadColumn.INFO_FILE,
                DownloadColumn.SUBTITLE,
                DownloadColumn.INTERRUPTED,
                    -> breite[column] = 0

                else -> Unit
            }
        }
    }

    private inner class TableRowTransferHandlerDownload(
        private val table: JTable,
    ) : TransferHandler() {
        private val localObjectFlavor = DataFlavor(Int::class.javaObjectType, "Integer Row Index")
        private var transferredRows: IntArray? = null

        override fun createTransferable(component: JComponent): Transferable {
            check(component == table)
            transferredRows = table.selectedRows
            return DataHandler(table.selectedRow, localObjectFlavor.mimeType)
        }

        override fun canImport(info: TransferSupport): Boolean {
            return try {
                val canImport = info.component == table && info.isDrop && info.isDataFlavorSupported(localObjectFlavor)
                table.cursor = if (canImport) DragSource.DefaultMoveDrop else DragSource.DefaultMoveNoDrop
                canImport
            } catch (exception: Exception) {
                logger.error("canImport", exception)
                true
            }
        }

        override fun getSourceActions(component: JComponent): Int = COPY_OR_MOVE

        override fun importData(info: TransferSupport): Boolean {
            return try {
                val target = info.component as JTable
                val dropLocation = info.dropLocation as TableDropLocation
                var index = dropLocation.row
                val max = table.model.rowCount
                if (index !in 0..max) {
                    index = max
                }
                target.cursor = Cursor.getPredefinedCursor(Cursor.DEFAULT_CURSOR)

                val rows = transferredRows ?: return false
                reorder(index, rows)
                transferredRows = null
                true
            } catch (exception: Exception) {
                logger.error("importData", exception)
                false
            }
        }

        private fun reorder(targetIndex: Int, rowsFrom: IntArray) {
            saveSelectedTableRows()

            val tableModel = model as TModelDownload

            sortDownloadListByTableRows()

            var insertionIndex = targetIndex
            val downloadsToMove = ArrayList<DatenDownload>()
            for (row in rowsFrom) {
                if (insertionIndex > row) {
                    --insertionIndex
                }

                val download = tableModel.getValueAt(convertRowIndexToModel(row), DownloadColumn.REF.index) as DatenDownload
                downloadsToMove.add(download)
            }

            downloads.moveDownloadsTo(insertionIndex, downloadsToMove)
            rowSorter?.sortKeys = null
            restoreSelectedTableRows()
        }

        override fun exportDone(source: JComponent, data: Transferable, action: Int) {
            if (action == MOVE) {
                table.cursor = Cursor.getPredefinedCursor(Cursor.DEFAULT_CURSOR)
            }
        }
    }

    private class DownloadsRowSorter(model: TableModel) : TriStateTableRowSorter<TableModel>(model) {
        override fun setModel(model: TableModel) {
            super.setModel(model)
            configureSortableColumns()
        }

        override fun setSortKeys(sortKeys: MutableList<out SortKey>?) {
            super.setSortKeys(sortKeys?.filter { isSortable(it.column) })
        }

        private fun configureSortableColumns() {
            setSortable(DownloadColumn.BUTTON_START.index, false)
            setSortable(DownloadColumn.BUTTON_DELETE.index, false)
        }
    }

}
